File: array_c_plus_plus/4.py
# coding=UTF-8
'''
@Author: httermin
@Date: 2019-12-23 21:30:29
'''
import sys
import os

sys.path.append("..")
class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        if len(nums2) < len(nums1):
            nums1, nums2 = nums2, nums1

        m, n = len(nums1), len(nums2)

        if m == 0 and n == 0:
            return -1
        elif m == 0:
            if n % 2:
                return nums2[n//2]
            else:
                return (nums2[n//2] + nums2[n//2 - 1]) / 2

        imax, imin = m, 0

        is_odd = (m + n) % 2

        while imin <= imax:
            i = (imax + imin) // 2
            j = ((m + n) // 2) - i 

            if (i < m) and nums1[i] < nums2[j-1]:
                imin += 1
            elif (i > 0) and nums1[i-1] > nums2[j]:
                imax -= 1
            else:
                if i == 0:
                    max_left = nums2[j-1]
                    if j != n:
                        min_right = min(nums1[i], nums2[j])
                    else:
                        min_right = nums1[i]
                elif i == m:
                    min_right = nums2[j]
                    if j != 0:
                        max_left = max(nums1[i-1], nums2[j-1])
                    else:
                        max_left = nums1[i-1]
                
                elif nums1[i] >= nums2[j-1] and nums1[i-1] <= nums2[j]:
                    max_left = max(nums1[i-1], nums2[j-1])
                    min_right = min(nums1[i], nums2[j])
                else:
                    max_left = -1
                    min_right = -1

                if is_odd:
                    return min_right
                else:
                    return (min_right + max_left) / 2
    
                
                
if __name__ == "__main__":
    ss = Solution()

    ex = [([2], [1, 3]), ([7, 8], [9, 10]), ([7, 8], [6, 9, 10]), ([7, 8, 11], [6, 9, 10]),
        ([11, 12], [6, 9, 10, 11]), ([6, 9, 10], [7, 8]), ([0,0], [0,0]), ([], [1,2,3])]


    for num1, num2 in ex:
        print(ss.findMedianSortedArrays(num1, num2))

File: array_c_plus_plus/4.cpp
/*
 * @Author: httermin
 * @Date: 2019-09-10 17:56:42
 */
#include <vector>
#include<iostream>
#include <algorithm>
#include<unordered_map>

using namespace std;

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int i = floor(nums1.size()/2) - 1;
        bool stop = false;
        int j = floor((nums1.size() + nums2.size())/2) - i - 2;

        double result;
        while(!stop){
            if(j + 1 < nums2.size() && nums1[i] > nums2[j+1]){
                i--;
                j = max(nums2.size(), floor((nums1.size() + nums2.size())/2) - i - 2);
            }else if{i + 1 < nums1.size() && nums1[i + 1] < nums2[j]}{
                i++;
                j = min(0, floor((nums1.size() + nums2.size())/2) - i - 2);
            }else{
                stop = true;
                result = (i + j + 2) / 2;
            }
        }
        return result;
    }
};
